### ISSUE TITLE
KeyError in synthesis core search

### ISSUE THREAD

**Issue by seizethedave** [uwplse/cozy#8]
`$ cozy specs/algebra.ds  --java f.java`

```
STARTING IMPROVEMENT JOB _query33671 (|examples|=0)
Traceback (most recent call last):
  File "/Users/dgrant/Dev/cozy/cozy/cozy/jobs.py", line 29, in _run
    self.run()
  File "/Users/dgrant/Dev/cozy/cozy/cozy/synthesis/high_level_interface.py", line 111, in run
    stop_callback=lambda: self.stop_requested)):
  File "/Users/dgrant/Dev/cozy/cozy/cozy/synthesis/core.py", line 603, in improve
    old_e, new_e, repl = learner.next()
  File "/Users/dgrant/Dev/cozy/cozy/cozy/synthesis/core.py", line 341, in next
    improvements = list(self._possible_replacements(*self.backlog))
  File "/Users/dgrant/Dev/cozy/cozy/cozy/synthesis/core.py", line 324, in _possible_replacements
    for (watched_e, r, watched_cost, assumptions, p, bound) in self.behavior_index.search(fp):
  File "/Users/dgrant/Dev/cozy/cozy/cozy/synthesis/core.py", line 161, in search
    yield from m[BehaviorIndex.VALUE]
KeyError: 'value'
```

**Comment by seizethedave** [comment]
@Calvin-L, this fixes the error but plz review. :)

**Comment by Calvin-L** [comment]
This is good defensive programming and we should leave it in.

The "right" fix is 9a835652, but it is highly non-intuitive. The tree of maps that `search` walks over has a precise known height (although the height is not passed to the `search` procedure). The KeyError here is because the calling code was expecting a different-height tree from `search`.
